### ISSUE TITLE
Tearing when using UV to access an array of sampler2d's

### ISSUE THREAD

**Issue by WaveyGames** [godotengine/godot#98288]
### Tested versions

Reproducible in: 4.3stable and 4.4dev3

### System information

Godot v4.3.stable - Windows 10.0.26100 - Vulkan (Forward+) - dedicated AMD Radeon RX 6600 (Advanced Micro Devices, Inc.; 32.0.12011.1036) - 12th Gen Intel(R) Core(TM) i3-12100F (8 Threads)

### Issue description

I wrote a short shader to have a mesh sample one image when UV.x < 0.5, and another image otherwise.
Here it is:
```
shader_type spatial;
uniform sampler2D[2] images : source_color;

void fragment() {
	ALBEDO = texture(images[int(step(0.5, UV.x))], UV).rgb;
}
```

But when you put it on a mesh and fill the array with two different images it looks something like this:
![tearing_showcase](https://github.com/user-attachments/assets/c18c718b-5847-41da-bfb1-5aa9ef7c3ba4)

I expected a clean line in the middle separating the two images, like this:
![image](https://github.com/user-attachments/assets/bf6f4bb6-c626-4d40-affb-9f1c9722803a)

The code for the above:
```
shader_type spatial;
uniform sampler2D[2] images : source_color;

vec3 get_color(int index, vec2 uv) {
	if(index == 1) {
		return texture(images[1], uv).rgb;
	}
	return texture(images[0], uv).rgb;
}
void fragment() {
	ALBEDO = get_color(int(step(0.5, UV.x)), UV);
}
```

### Steps to reproduce

After you open the MRP, the issue should be visible immediately.

### Minimal reproduction project (MRP)

[MRP_bug.zip](https://github.com/user-attachments/files/17430209/MRP_bug.zip)

**Comment by tetrapod00** [comment]
I can't reproduce it on 4.3. Possibly hardware or driver specific?

![Godot_v4 3-stable_win64_42RzYf4qGx](https://github.com/user-attachments/assets/88c60065-04da-4f83-80b3-a88e213193a3)

Godot v4.3.stable - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1660 Ti with Max-Q Design (NVIDIA; 32.0.15.5612) - AMD Ryzen 7 3750H with Radeon Vega Mobile Gfx (8 Threads)

**Comment by DarioSamo** [comment]
This is a pretty typical artifact from not indicating the access is not uniform explicitly in the shader.

https://anki3d.org/resource-uniformity-bindless-access-in-vulkan/

> One additional piece to the puzzle is the NonUniform SPIR-V decoration which is exposed via nonuniformEXT in GLSL and NonUniformResourceIndex() in HLSL.

I don't believe access to this modifier is exposed in GDShader. It might take a while to get a feature like that added and it's a pretty highly technical modifier to expect people to know about it, so my suggestion would be you attempt some other way to do this. Try just sampling both of the images and mixing based on the step UV value.

**Comment by DarioSamo** [comment]
> I can't reproduce it on 4.3. Possibly hardware or driver specific?

Yes, AMD is more prone to showing artifacts when non uniform access is not specified.

**Comment by stuartcarnie** [comment]
Is there a downside to having the Godot shader compiler always insert the `nonuniformEXT` decorator for array indexing of textures? According to the article:

> If an implementation doesn’t care then it will simply ignore it [`nonuniformEXT`].

**Comment by WaveyGames** [comment]
> My suggestion would be you attempt some other way to do this. Try just sampling both of the images and mixing based on the step UV value. That should guarantee the access to both samplers is uniform.

I found a way around the issue using a bunch of step functions, since the original problem is sampling from a couple more textures, and the result looks great! Thanks!
